### PR TITLE
Sharding update

### PR DESCRIPTION
**Pull request by zbynek001**
multiple sharding updates

* Log non-completed rebalance at warning level
migrated from https://github.com/akka/akka/pull/28335

* Shard region registration to more potential oldest
migrated from https://github.com/akka/akka/pull/28470

* Improve error message when coordinator not found
migrated from https://github.com/akka/akka/pull/28576

* Don't initialize durable keys when rememberEntities not used
migrated from https://github.com/akka/akka/pull/28568

* Use dedicated message for PersistentShardCoodinator termination instead of PoisonPill
migrated from https://github.com/akka/akka/pull/28104

* Reconsider cluster.role.<role-name>.min-nr-of-members fallback
migrated from https://github.com/akka/akka/pull/28203

* Avoid dead letter for rebalance timeout msg
migrated from https://github.com/akka/akka/pull/28274

### CODE REVIEW

**Comment by Aaronontheweb** [comment]
@zbynek001 I'll review this tonight.

File: src/contrib/cluster/Akka.Cluster.Sharding.Tests.MultiNode/ClusterShardingGracefulShutdownSpec.cs [hunk 1]
```diff
@@ -82,7 +82,7 @@ public class PersistentClusterShardingGracefulShutdownSpec : ClusterShardingGrac
     public class DDataClusterShardingGracefulShutdownSpec : ClusterShardingGracefulShutdownSpec
     {
         public DDataClusterShardingGracefulShutdownSpec() : this(new DDataClusterShardingGracefulShutdownSpecConfig()) { }
-        protected DDataClusterShardingGracefulShutdownSpec(DDataClusterShardingGracefulShutdownSpecConfig config) : base(config, typeof(PersistentClusterShardingGracefulShutdownSpec)) { }
+        protected DDataClusterShardingGracefulShutdownSpec(DDataClusterShardingGracefulShutdownSpecConfig config) : base(config, typeof(DDataClusterShardingGracefulShutdownSpec)) { }
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.Cluster.Sharding.Tests.MultiNode/ClusterShardingGracefulShutdownSpec.cs]: Nice catch

File: src/contrib/cluster/Akka.Cluster.Sharding/ClusterSharding.cs [hunk 2]
```diff
@@ -1290,7 +1292,7 @@ public RebalanceWorker(string shard, IActorRef @from, TimeSpan handOffTimeout, I
             foreach (var region in _remaining)
                 region.Tell(new PersistentShardCoordinator.BeginHandOff(shard));
 
-            Context.System.Scheduler.ScheduleTellOnce(handOffTimeout, Self, ReceiveTimeout.Instance, Self);
+            Timers.StartSingleTimer("hand-off-timeout", ReceiveTimeout.Instance, handOffTimeout);
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.Cluster.Sharding/ClusterSharding.cs]: LGTM

File: src/contrib/cluster/Akka.Cluster.Sharding/ClusterShardingGuardian.cs [hunk 3]
```diff
@@ -240,6 +249,16 @@ public ClusterShardingGuardian()
             });
         }
 
+        private ReplicatorSettings GetReplicatorSettings(ClusterShardingSettings shardingSettings)
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.Cluster.Sharding/ClusterShardingGuardian.cs]: LGTM

File: src/contrib/cluster/Akka.Cluster.Sharding/DDataShardCoordinator.cs [hunk 4]
```diff
@@ -203,8 +213,18 @@ private bool WaitingForStateInitialized(object message)
...
+                        Log.Error("The ShardCoordinator was unable to update a distributed state within 'updating-state-timeout': {0} ({1}), event={2}", _writeConsistency.Timeout, _terminating ? "terminating" : "retrying", e);
+
+                        if (_terminating)
+                        {
+                            Context.Stop(Self);
+                        }
+                        else
+                        {
+                            // repeat until UpdateSuccess
+                            SendCoordinatorStateUpdate(e);
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.Cluster.Sharding/DDataShardCoordinator.cs]: LGTM

File: src/contrib/cluster/Akka.Cluster.Sharding/ShardRegion.cs [hunk 5]
```diff
@@ -328,6 +328,9 @@ internal static Props ProxyProps(string typeName, ClusterShardingSettings settin
         /// </summary>
         protected IImmutableSet<Member> MembersByAge = ImmutableSortedSet<Member>.Empty.WithComparer(AgeOrdering);
 
+        // membersByAge contains members with these status
+        private static readonly HashSet<MemberStatus> MemberStatusOfInterest = new HashSet<MemberStatus> { MemberStatus.Up, MemberStatus.Leaving, MemberStatus.Exiting };
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.Cluster.Sharding/ShardRegion.cs]: If you want this to be read-only, change it to an `ImmutableHashSet` or some other non-modifiable object.

File: src/contrib/cluster/Akka.Cluster.Sharding/ShardRegion.cs [hunk 6]
```diff
@@ -557,21 +573,29 @@ private void InitializeShard(ShardId id, IActorRef shardRef)
 
         private void Register()
         {
-            var coordinator = CoordinatorSelection;
-            coordinator?.Tell(RegistrationMessage);
+            var actorSelections = CoordinatorSelection;
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.Cluster.Sharding/ShardRegion.cs]: Makes sense - look for the coordinator in multiple locations if needed

File: src/contrib/cluster/Akka.Cluster.Sharding/ShardingMessages.cs [hunk 7]
```diff
@@ -22,6 +23,18 @@ public interface IShardRegionCommand { }
     /// </summary>
     public interface IShardRegionQuery { }
 
+    /// <summary>
+    /// Used as a special termination message from <see cref="ClusterSharding"/>
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.Cluster.Sharding/ShardingMessages.cs]: Might want to specify that this is only for `Shard` actors

**Comment by Aaronontheweb** [comment]
@zbynek001 left some minor change requests for you

File: src/contrib/cluster/Akka.Cluster.Sharding/ShardingMessages.cs (hunk 7 above)
Comment (zbynek001) [reply src/contrib/cluster/Akka.Cluster.Sharding/ShardingMessages.cs]: changed it slightly

**Pull request merged**